---
title: "Proportion Test"
author: "Yaamini Venkataraman"
date: "9/15/2019"
output: html_document
---

I will use a proportion test to compare the proportion of genome feature overlaps between differentially methylated loci (DML) and the gene background. I will also examine differences in genomic location bewteen *C. gigas* and previously-analyzed *C. virginica* DML.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load packages

```{r}
#install.packages("dichromat") #Install package for looking at colors from colorblind perspective
require("dichromat") #Load package
```

# Obtain session information

```{r}
sessionInfo()
```

# *C. gigas* only

## Import data

```{r}
overlapData <- read.csv("2019-09-15-Overlap-Proportions.csv", header = TRUE)
rownames(overlapData) <- overlapData$genomicFeature #Set genomic feature indication and rownames
overlapData <- overlapData[,-1] #Remove genomic feature indication column
head(overlapData) #Confirm import
```

## Reformat data

```{r}
proportionData <- overlapData #Copy overlap data as a new dataframe
proportionData <- proportionData[-6,] #Remove totalLines row
totalEntries <- c(sum(proportionData$totalCpG),
                  sum(proportionData$DML)) #Sum the number of CpGs in each category
proportionData <- rbind(proportionData, totalEntries) #Add totalEntries to bottom of proportionData
row.names(proportionData)[6] <- "totalEntries" #Add rowname
head(proportionData) #Confirm changes
```

```{r}
nLength <- length(proportionData$totalCpG) #Count number of rows
for(i in 1:nLength) {
  proportionData[i,] <- (proportionData[i,]/proportionData[6,])*100
} #Divide each column of proportionData by respective totalEntries entry. Multiply by 100 and and save the percentage
proportionData <- proportionData[-6,] #Remove totalEntries
head(proportionData) #Confirm changes
```

## Conduct chi-squared tests of homogeneity

The null hypothesis is that loci distributions in the genome are the same between different categories. Even though the total CpGs are not the background of the DML, it's still an interesting comparison.

```{r}
methylatedVersusDMLTest <- chisq.test(t(proportionData)) #Conduct a chi-squared test
methylatedVersusDMLTest #The distribution of DML is significantly different from CpGs.
```

## Create figures

```{r}
plotColors <- rev(RColorBrewer::brewer.pal(5, "RdPu")) #Create a color palette for the barplots. Use 5 purple shades from RColorBrewer. Reverse the order so the darkest shade is used first.
barplot(t(t(proportionData)),
        col = dichromat(plotColors)) #Check that plot colors will be easy to interpret for those with color blindness
```

```{r}
#pdf("2019-09-15-Total-CpGs-Versus-DML.pdf", height = 8.5, width = 11)

par(mar = c(3,5,1,1), oma = c(1, 1, 1, 10)) #Change figure boundaries
barplot(t(t(proportionData)), 
        axes = FALSE, 
        names.arg = c("All CpGs", "DML"),
        ylim = c(0,100),
        col = plotColors) #Create a stacked barplot. Use axes = FALSE to remove the y-axis and names.arg to set labels on the x-axis.
axis(side = 2, at = seq(0, 100, by = 10), las = 2, col = "grey80")
mtext(side = 2, "Proportion CpGs", line = 3) #Add y-axis label

par(fig = c(0, 1, 0, 1), oma = c(0, 0, 0, 0), mar = c(0, 0, 0, 0), new = TRUE) #Create new plot
plot(0, 0, type = "n", bty = "n", xaxt = "n", yaxt = "n") #Add new plot on top of current plot
legend("topright",
       xpd = TRUE,
       legend = c("Exons", "Introns", "Transposable Elements", "Putative Promoters", "Other"),
       pch = 22, 
        col = "black", 
        pt.bg = plotColors,
       bty = "n") #Place a legend in the top right of the figure with no box

#dev.off()
```

# *C. gigas* vs. *C. virginica*

Information for the *C. virginica* DML can be found [here](https://github.com/epigeneticstoocean/paper-gonad-meth).

## Import data

```{r}
comboData <- read.csv("2020-02-12-Combined-Overlap-Proportions.csv", header = TRUE)
rownames(comboData) <- comboData$genomicFeature #Set genomic feature indication and rownames
comboData <- comboData[,-1] #Remove genomic feature indication column
head(comboData) #Confirm import
```

## Reformat data

```{r}
comboPropData <- comboData #Copy overlap data as a new dataframe
comboPropData <- comboPropData[-6,] #Remove totalLines row
comboTotalEntries <- c(sum(comboPropData$Cvirg),
                  sum(comboPropData$Cgigas)) #Sum the number of CpGs in each category
comboPropData <- rbind(comboPropData, comboTotalEntries) #Add comboTotalEntries to bottom of proportionData
row.names(comboPropData)[6] <- "totalEntries" #Add rowname
head(comboPropData) #Confirm changes
```

```{r}
nLength <- length(comboPropData$Cvirg) #Count number of rows
for(i in 1:nLength) {
  comboPropData[i,] <- (comboPropData[i,]/comboPropData[6,])*100
} #Divide each column of comboPropData by respective totalEntries entry. Multiply by 100 and and save the percentage
comboPropData <- comboPropData[-6,] #Remove totalEntries
head(comboPropData) #Confirm changes
```

## Conduct chi-squared tests of homogeneity

The null hypothesis is that loci distributions amongst various genomic features are the same in both species.

```{r}
speciesDMLTest <- chisq.test(t(comboPropData)) #Conduct a chi-squared test
speciesDMLTest #The distribution of DML is significantly different between species.
```

## Create figures

```{r}
RColorBrewer::display.brewer.all(colorblindFriendly = TRUE) #Only display colorblind-friendly palettes
```


```{r}
plotColors2 <- rev(RColorBrewer::brewer.pal(5, "Oranges")) #Create a color palette for the barplots. Use 5 purple shades from RColorBrewer. Reverse the order so the darkest shade is used first.
barplot(t(t(comboPropData)),
        col = dichromat(plotColors2)) #Check that plot colors will be easy to interpret for those with color blindness
```

```{r}
#pdf("2020-02-12-Cvirg-Versus-Cgigas-DML.pdf", height = 8.5, width = 11)

par(mar = c(3,5,1,1), oma = c(1, 1, 1, 10)) #Change figure boundaries
barplot(t(t(comboPropData)), 
        axes = FALSE, 
        names.arg = c(expression(italic("C. virginica")), expression(italic("C. gigas"))),
        ylim = c(0,100),
        col = plotColors2) #Create a stacked barplot. Use axes = FALSE to remove the y-axis and names.arg to set labels on the x-axis. Use expression(italic()) to italicize species names
axis(side = 2, at = seq(0, 100, by = 10), las = 2, col = "grey80")
mtext(side = 2, "Proportion CpGs", line = 3) #Add y-axis label

par(fig = c(0, 1, 0, 1), oma = c(0, 0, 0, 0), mar = c(0, 0, 0, 0), new = TRUE) #Create new plot
plot(0, 0, type = "n", bty = "n", xaxt = "n", yaxt = "n") #Add new plot on top of current plot
legend("topright",
       xpd = TRUE,
       legend = c("Putative Promoters", "Exons", "Introns", "Transposable Elements", "Other"),
       pch = 22, 
        col = "black", 
        pt.bg = plotColors2,
       bty = "n") #Place a legend in the top right of the figure with no box

#dev.off()
```


