---
title: "Identification of DML"
author: "Yaamini Venkataraman"
date: "03/10/2021"
output: github_document
---

In this script, I'll get preliminary methylation information from oyster samples. Additionally, identify differentially methylated loci (DML) between pH treatments. I have any covariate information for sex and stage, but not tank, shell height, width, or wet weight because the some samples were mixed up during histology processing.

# Prepare R Markdown file

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/Users/yaamini/Documents/project-gigas-oa-meth/output/06-methylKit/") #Set root directory
```

# Install packages

```{r}
#Packages for running methylKit

#install.packages("devtools") #Install the devtools package
require(devtools) #Load devtools
#source("http://bioconductor.org/biocLite.R") #Source package from bioconductor
#biocLite("methylKit") #Install methylkit
#install_github("al2na/methylKit", build_vignettes = FALSE, repos = BiocManager::repositories(), dependencies = TRUE) #Install more methylKit options
require(methylKit) #Load methylkit
#install.packages("tidyverse")
require(tidyverse)
#install.packages("dplyr")
require(dplyr)
```

```{r}
#Packages for visualization

#install.packages("vegan")
#install.packages("gplots")
#install.packages("dichromat")

require(vegan)
require(gplots)
require(dichromat)
```

# Obtain session information

```{r}
sessionInfo()
```

# Download files from `gannet`

```{bash}
wget -r -l1 --no-parent --no-directories -A.merged_CpG_evidence.cov https://gannet.fish.washington.edu/spartina/project-gigas-oa-meth/output/bismark-roslin/ #Download files from gannet to the current directory (where script is saved). Exclude directory structure
```

```{bash}
mkdir ../output/06-methylKit #Make methylkit output directory
```

```{bash}
mv *.merged_CpG_evidence.cov ../output/06-methylKit #Move files from current directory to analyses folder
```

## Process methylation data

```{r}
analysisFiles <- list("zr3616_1_R1_val_1_val_1_val_1_bismark_bt2_pe..CpG_report.merged_CpG_evidence.cov",
                      "zr3616_2_R1_val_1_val_1_val_1_bismark_bt2_pe..CpG_report.merged_CpG_evidence.cov",
                      "zr3616_3_R1_val_1_val_1_val_1_bismark_bt2_pe..CpG_report.merged_CpG_evidence.cov",
                      "zr3616_4_R1_val_1_val_1_val_1_bismark_bt2_pe..CpG_report.merged_CpG_evidence.cov",
                      "zr3616_5_R1_val_1_val_1_val_1_bismark_bt2_pe..CpG_report.merged_CpG_evidence.cov",
                      "zr3616_6_R1_val_1_val_1_val_1_bismark_bt2_pe..CpG_report.merged_CpG_evidence.cov",
                      "zr3616_7_R1_val_1_val_1_val_1_bismark_bt2_pe..CpG_report.merged_CpG_evidence.cov",
                      "zr3616_8_R1_val_1_val_1_val_1_bismark_bt2_pe..CpG_report.merged_CpG_evidence.cov") #Put all .cov files into a list for analysis.
```

```{r}
sampleMetadata <- data.frame("sampleID" = c("1", "2", "3", "4", "5", "6", "7", "8"),
                             "slide-label" = c("02-T1", "04-T3", "06-T1", "08-T2", "11-T4", "UK-02", "UK-04", "UK-06"),
                             "pHTreatment" = c(rep(1, times = 4), rep(0, times = 4))) #Create dataframe with metadata. 1 = low, 0 = ambient. Slide label information from histology processing
head(sampleMetadata) #Confirm dataframe creation
```

I'll use `methRead` to create a methylation object from the coverage files, and include sample ID and treatment information.

```{r}
processedFiles <- methylKit::methRead(analysisFiles,
                                      sample.id = list("1", "2", "3", "4",
                                                       "5", "6", "7", "8"),
                                      assembly = "Roslin",
                                      treatment = sampleMetadata$pHTreatment,
                                      pipeline = "bismarkCoverage",
                                      mincov = 2) #Process files. Treatment specified based on ploidy status. Use mincov = 2 to quickly process reads.
```

```{r}
processedFilteredFilesCov5 <- methylKit::filterByCoverage(processedFiles,
                                                          lo.count = 5, lo.perc = NULL,
                                                          high.count = NULL, high.perc = 99.9) %>%
  methylKit::normalizeCoverage(.) #Filter coverage information for minimum 5x coverage, and remove PCR duplicates by excluding data in the 99.9th percentile of coverage with hi.perc = 99.9. Normalize coverage between samples to avoid over-sampling reads from one sample during statistical testing
```

```{r}
save.image("methylKit.RData") #Save R Data in case R crashes
#load("methylKit.RData") #Load R Data
```

# Characterize general methylation

## Sample-specific descriptive statistics

### Specify working directory for output

```{bash}
mkdir general-stats
```

```{r}
nFiles <- 8 #Count number of samples
fileName <- data.frame("nameBase" = rep("general-stats/percent-CpG-methylation", times = nFiles),
                       "nameBase2" = rep("general-stats/percent-CpG-coverage", times = nFiles),
                       "sample.ID" = c(1:8)) #Create new dataframe for filenames.
head(fileName) #Confirm dataframe creation
```

```{r}
fileName$actualFileName1 <- paste(fileName$nameBase, "-Filtered", "-5xCoverage", "-Sample", fileName$sample.ID, ".jpeg", sep = "") #Create a new column for the full filename for filtered + 5x coverage + specific sample's percent CpG methylation plot
fileName$actualFileName2 <- paste(fileName$nameBase2, "-Filtered", "-5xCoverage", "-Sample", fileName$sample.ID, ".jpeg", sep = "") #Create a new column for the full filename for filtered + 5x coverage + specific sample's percent CpG coverage plot
head(fileName) #Confirm column creation
```

### Create plots

```{r}
for(i in 1:nFiles) { #For each data file
  jpeg(filename = fileName$actualFileName1[i], height = 1000, width = 1000) #Save file with designated name
  methylKit::getMethylationStats(processedFilteredFilesCov5[[i]], plot = TRUE, both.strands = FALSE) #Get %CpG methylation information
  dev.off() #Turn off plotting device
} #Plot and save %CpG methylation information
```

```{r}
for(i in 1:nFiles) { #For each data file
  jpeg(filename = fileName$actualFileName2[i], height = 1000, width = 1000) #Save file with designated name
  methylKit::getCoverageStats(processedFilteredFilesCov5[[i]], plot = TRUE, both.strands = FALSE) #Get CpG coverage information
  dev.off() #Turn off plotting device
} #Plot and save CpG coverage information
```

## Comparative analysis

```{r}
methylationInformationFilteredCov5 <- methylKit::unite(processedFilteredFilesCov5, 
                                                       destrand = FALSE,
                                                       mc.cores = 2) #Combine all processed files into a single table. Use destrand = FALSE to not destrand. By default only bases with data in all samples will be kept
head(methylationInformationFilteredCov5) #Confirm unite
```

```{r}
save.image("methylKit.RData") #Save R Data in case R crashes
#load("methylKit.RData") #Load R Data
```

```{r}
clusteringInformationFilteredCov5 <- methylKit::clusterSamples(methylationInformationFilteredCov5, dist = "correlation", method = "ward", plot = FALSE) #Save cluster information as a new object
```

```{r}
jpeg(filename = "general-stats/Full-Sample-Pearson-Correlation-Plot-FilteredCov5Destrand.jpeg", height = 1000, width = 1000) #Save file with designated name
methylKit::getCorrelation(methylationInformationFilteredCov5, plot = TRUE) #Understand correlation between methylation patterns in different samples
dev.off()
```

```{r}
jpeg(filename = "general-stats/Full-Sample-CpG-Methylation-Clustering-FilteredCov5Destrand.jpeg", height = 1000, width = 1000) #Save file with designated name
methylKit::clusterSamples(methylationInformationFilteredCov5, dist = "correlation", method = "ward", plot = TRUE) #Cluster samples based on correlation coefficients
dev.off()
```

```{r}
jpeg(filename = "general-stats/Full-Sample-Methylation-PCA-FilteredCov5Destrand.jpeg", height = 1000, width = 1000) #Save file with designated name
methylKit::PCASamples(methylationInformationFilteredCov5) #Run a PCA analysis on percent methylation for all samples
dev.off() #Turn off plotting device
```

```{r}
jpeg(filename = "general-stats/Full-Sample-Methylation-Screeplot-FilteredCov5Destrand.jpeg", height = 1000, width = 1000) #Save file with designated name
methylKit::PCASamples(methylationInformationFilteredCov5, screeplot = TRUE) #Run the PCA analysis and plot variances against PC number in a screeplot
dev.off()
```

```{r}
save.image("methylKit.RData") #Save R Data in case R crashes
#load("methylKit.RData") #Load R Data
```

# Differentially methylated loci

I have three female samples and one indeterminate sample per treatment. Although I produced comparative plots for I will identify DML for these datasets separately.

## Female samples

I'll first separate the female samples from the full dataset, then conduct comparative analyses within this subset.

```{r}
methylationInformationFilteredCov5Fem <- methylKit::reorganize(methylationInformationFilteredCov5,
                                                               sample.ids = c("1", "3", "4",
                                                                              "6", "7", "8"),
                                                               treatment = c(rep(1, times = 3),
                                                                             rep(0, times = 3))) #Get female sample information from methylBase object created by unite by specifying sample IDs. Include treatment information as well
```

```{r}
jpeg(filename = "general-stats/Full-Sample-CpG-Methylation-Clustering-FilteredCov5Destrand-Fem.jpeg", height = 1000, width = 1000) #Save file with designated name
methylKit::clusterSamples(methylationInformationFilteredCov5Fem, dist = "correlation", method = "ward", plot = TRUE) #Cluster female samples based on correlation coefficients
dev.off()
```

```{r}
jpeg(filename = "general-stats/Full-Sample-Methylation-PCA-FilteredCov5Destrand-Fem.jpeg", height = 1000, width = 1000) #Save file with designated name
methylKit::PCASamples(methylationInformationFilteredCov5Fem) #Run a PCA analysis on percent methylation for female samples
dev.off() #Turn off plotting device
```

```{r}
jpeg(filename = "general-stats/Full-Sample-Methylation-Screeplot-FilteredCov5Destrand-Fem.jpeg", height = 1000, width = 1000) #Save file with designated name
methylKit::PCASamples(methylationInformationFilteredCov5Fem, screeplot = TRUE) #Run the PCA analysis and plot variances against PC number in a screeplot for female samples
dev.off()
```

### Create covariate matrix

I don't have covariate information (tank, wet weight, shell length) for all samples, since some tissues were mixed up during histology processing. However, I do have sex and stage information for each sample that will likely contribute to methylation patterns. Information is from [this spreadsheet](https://github.com/RobertsLab/project-oyster-oa/blob/master/data/Manchester/2018-02-27-Gigas-Histology-Classification.csv), and slide label information is consistent with `sampleMetadata`.

```{r}
covariateMetadataFem <- data.frame("stage" = c("2", "1", "1",
                                               "1", "2", "3")) #Create dataframe with stage information. 1/2/3 = early female maturation, with larger numbers indicating more maturity
head(covariateMetadataFem) #Check dataframe format
```

### Identify DML

```{r}
#Code that was used to test calculateDiffMeth parameters

#differentialMethylationStatsTreatment <- methylKit::calculateDiffMeth(methylationInformationFilteredCov5)
#differentialMethylationStatsTreatment2 <- methylKit::calculateDiffMeth(methylationInformationFilteredCov5, overdispersion = "MN", test = "Chisq")
#head(differentialMethylationStatsTreatment) #Look at differential methylation statistics
#head(differentialMethylationStatsTreatment2) #Look at differential methylation statistics
```

```{r}
differentialMethylationStatsTreatmentFem2 <- methylKit::calculateDiffMeth(methylationInformationFilteredCov5Fem, covariates = covariateMetadataFem, overdispersion = "MN", test = "Chisq") #Calculate differential methylation statistics and include covariate information.
head(differentialMethylationStatsTreatmentFem2) #Look at differential methylation statistics
```

```{r}
diffMethStatsTreatment25Fem <- methylKit::getMethylDiff(differentialMethylationStatsTreatmentFem2, difference = 25, qvalue = 0.01)
length(diffMethStatsTreatment25Fem$chr) #20830 DML
head(diffMethStatsTreatment25Fem)
```

```{r}
diffMethStatsTreatment75Fem <- methylKit::getMethylDiff(differentialMethylationStatsTreatmentFem2, difference = 75, qvalue = 0.01)
length(diffMethStatsTreatment75Fem$chr) #315 DML
head(diffMethStatsTreatment75Fem)
```

```{r}
save.image("methylKit.RData") #Save R Data in case R crashes
#load("methylKit.RData") #Load R Data
```

```{r}
write.csv(diffMethStatsTreatment25Fem, "DML/DML-pH-25-Cov5-Fem.csv")
write.csv(diffMethStatsTreatment50Fem, "DML/DML-pH-50-Cov5-Fem.csv")
write.csv(diffMethStatsTreatment75Fem, "DML/DML-pH-75-Cov5-Fem.csv")
```

### Randomization trial

```{r}
# Code used to test initial randomization trials
# for (i in 1:1000) { #For each iteration
#   pHRandTreat <- sample(sampleMetadata$pHTreatment, 
#                         size = length(sampleMetadata$pHTreatment),
#                         replace = FALSE) #Randomize treatment information
#   pHRandDML <- methylKit::reorganize(processedFiles,
#                                      sample.ids = sampleMetadata$sampleID,
#                                      treatment = pHRandTreat) %>%
#     methylKit::filterByCoverage(., lo.count = 5, lo.perc = NULL, 
#                                 hi.count = NULL, hi.perc = 99.9) %>%
#     methylKit::normalizeCoverage(.) %>%
#     methylKit::unite(., destrand = FALSE) %>%
#     methylKit::calculateDiffMeth(.) %>%
#     methylKit::getMethylDiff(., difference = 50, qvalue = 0.01) %>% 
#     nrow() -> pHRandTest25[i] #Reorganize, filter, normalize, and unite samples. Calculate diffMeth and obtain DML that are 50% and have a q-value of 0.01. Save the number of DML identified
# }
```

```{r}
sampleMetadataFem <- data.frame("sampleID" = c("1", "3", "4",
                                               "6", "7", "8"),
                             "slide-label" = c("02-T1", "06-T1", "08-T2",
                                               "UK-02", "UK-04", "UK-06"),
                             "pHTreatment" = c(rep(1, times = 3),
                                               rep(0, times = 3))) #Create dataframe with metadata for female samples. 1 = low, 0 = ambient. Slide label information from histology processing
head(sampleMetadataFem) #Confirm dataframe creation
```

#### 75% difference

```{r}
pHRandTest75Fem <- NULL #Create an empty matrix to store randomization trial results for 25% difference
```

```{r}
for (i in 1:1000) { #For each iteration
  pHRandTreat <- sample(sampleMetadataFem$pHTreatment, 
                        size = length(sampleMetadataFem$pHTreatment),
                        replace = FALSE) #Randomize female treatment information
  pHRandDML <- methylKit::reorganize(methylationInformationFilteredCov5Fem,
                                     sample.ids = sampleMetadataFem$sampleID,
                                     treatment = pHRandTreat) %>%
    methylKit::calculateDiffMeth(., covariates = covariateMetadataFem, 
                                 overdispersion = "MN", 
                                 test = "Chisq") %>%
    methylKit::getMethylDiff(., difference = 75, qvalue = 0.01) %>%
    nrow() -> pHRandTest75Fem[i] #Shuffle treatments from methylBase object created by unite. Calculate diffMeth and obtain DML that are 75% and have a q-value of 0.01. Save the number of DML identified
}
```

```{r}
pHRandTestResults75Fem <- t.test(pHRandTest75Fem, alternative = "less", mu = nrow(diffMethStatsTreatment75Fem)) #Conduct a one-sample t-test to determine the probability of identifying more DML by chance
pHRandTestResults75Fem$statistic #t
pHRandTestResults75Fem$parameter #df
pHRandTestResults75Fem$p.value #P-value
```

```{r}
jpeg(filename = "../output/06-methylKit/rand-test/Random-Distribution-diff75-Fem.jpeg", height = 1000, width = 1000) #Save file with designated name
hist(pHRandTest75Fem, plot = TRUE, main = "", xlab = "Number of Female-DML (75% difference)") #Plot a histogram with randomization trial results
abline(v = nrow(diffMethStatsTreatment75Fem)) #Add a vertical line for the number of DML identified with getMethylDiff
dev.off()
```

```{r}
save.image("../output/06-methylKit/methylKit.RData") #Save R Data in case R crashes
#load("../output/06-methylKit/methylKit.RData") #Load R Data
```

## Indeterminate samples

```{r}
methylationInformationFilteredCov5Ind <- methylKit::reorganize(methylationInformationFilteredCov5,
                                                               sample.ids = c("2",
                                                                              "5"),
                                                               treatment = c(1, 0)) #Get indeterminate sample information from methylBase object created by unite by specifying sample IDs. Include treatment information as well
```

```{r}
jpeg(filename = "general-stats/Full-Sample-CpG-Methylation-Clustering-FilteredCov5Destrand-Ind.jpeg", height = 1000, width = 1000) #Save file with designated name
methylKit::clusterSamples(methylationInformationFilteredCov5Ind, dist = "correlation", method = "ward", plot = TRUE) #Cluster indeterminate samples based on correlation coefficients
dev.off()
```

```{r}
jpeg(filename = "general-stats/Full-Sample-Methylation-PCA-FilteredCov5Destrand-Ind.jpeg", height = 1000, width = 1000) #Save file with designated name
methylKit::PCASamples(methylationInformationFilteredCov5Ind) #Run a PCA analysis on percent methylation for indeterminate samples
dev.off() #Turn off plotting device
```

```{r}
jpeg(filename = "general-stats/Full-Sample-Methylation-Screeplot-FilteredCov5Destrand-Ind.jpeg", height = 1000, width = 1000) #Save file with designated name
methylKit::PCASamples(methylationInformationFilteredCov5Ind, screeplot = TRUE) #Run the PCA analysis and plot variances against PC number in a screeplot for indeterminate samples
dev.off()
```

### Identify DML

```{r}
differentialMethylationStatsTreatmentInd <- methylKit::calculateDiffMeth(methylationInformationFilteredCov5Ind) #Calculate differential methylation statistics.
head(differentialMethylationStatsTreatment) #Look at differential methylation statistics
```

```{r}
diffMethStatsTreatment25Ind <- methylKit::getMethylDiff(differentialMethylationStatsTreatmentInd, difference = 25, qvalue = 0.01)
length(diffMethStatsTreatment25Ind$chr) #129040
head(diffMethStatsTreatment25Ind)
```

```{r}
diffMethStatsTreatment50Ind <- methylKit::getMethylDiff(differentialMethylationStatsTreatmentInd, difference = 50, qvalue = 0.01)
length(diffMethStatsTreatment50Ind$chr) #73414
head(diffMethStatsTreatment50Ind)
```

```{r}
diffMethStatsTreatment75Ind <- methylKit::getMethylDiff(differentialMethylationStatsTreatmentInd, difference = 75, qvalue = 0.01)
length(diffMethStatsTreatment75Ind$chr) #21891
head(diffMethStatsTreatment75Ind)
```

```{r}
diffMethStatsTreatment100Ind <- methylKit::getMethylDiff(differentialMethylationStatsTreatmentInd, difference = 99, qvalue = 0.01)
length(diffMethStatsTreatment100Ind$chr) #2861
head(diffMethStatsTreatment100Ind)
```

```{r}
save.image("methylKit.RData") #Save R Data in case R crashes
#load("methylKit.RData") #Load R Data
```

```{r}
write.csv(diffMethStatsTreatment25Ind, "DML/DML-pH-25-Cov5-Ind.csv")
write.csv(diffMethStatsTreatment50Ind, "DML/DML-pH-50-Cov5-Ind.csv")
write.csv(diffMethStatsTreatment75Ind, "DML/DML-pH-75-Cov5-Ind.csv")
write.csv(diffMethStatsTreatment100Ind, "DML/DML-pH-100-Cov5-Ind.csv")
```

### Randomization trial

```{r}
sampleMetadataInd <- data.frame("sampleID" = c("2",
                                               "5"),
                             "slide-label" = c("04-T3",
                                               "11-T4"),
                             "pHTreatment" = c(1,
                                               0)) #Create dataframe with metadata for indeterminate samples. 1 = low, 0 = ambient. Slide label information from histology processing
head(sampleMetadataInd) #Confirm dataframe creation
```

#### 99% difference

```{r}
pHRandTest99Ind <- NULL #Create an empty matrix to store randomization trial results for 99% difference
```

```{r}
for (i in 1:1000) { #For each iteration
  pHRandTreat <- sample(sampleMetadataInd$pHTreatment, 
                        size = length(sampleMetadataInd$pHTreatment),
                        replace = FALSE) #Randomize indeterminate treatment information
  pHRandDML <- methylKit::reorganize(methylationInformationFilteredCov5Ind,
                                     sample.ids = sampleMetadataInd$sampleID,
                                     treatment = pHRandTreat) %>%
    methylKit::calculateDiffMeth(.) %>%
    methylKit::getMethylDiff(., difference = 99, qvalue = 0.01) %>%
    nrow() -> pHRandTest99Ind[i] #Shuffle treatments from methylBase object created by unite. Calculate diffMeth and obtain DML that are 99% and have a q-value of 0.01. Save the number of DML identified
}
```

```{r}
pHRandTestResults99Ind <- t.test(pHRandTest99Ind, alternative = "less", mu = nrow(diffMethStatsTreatment99Ind)) #Conduct a one-sample t-test to determine the probability of identifying more DML by chance
pHRandTestResults99Ind$statistic #t
pHRandTestResults99Ind$parameter #df
pHRandTestResults99Ind$p.value #P-value
```

```{r}
jpeg(filename = "rand-test/Random-Distribution-diff99-Ind.jpeg", height = 1000, width = 1000) #Save file with designated name
hist(pHRandTest99Ind, plot = TRUE, main = "", xlab = "Number of Indeterminate-DML (99% difference)") #Plot a histogram with randomization trial results
abline(v = nrow(diffMethStatsTreatment99Ind)) #Add a vertical line for the number of DML identified with getMethylDiff
dev.off()
```

```{r}
save.image("methylKit.RData") #Save R Data in case R crashes
#load("methylKit.RData") #Load R Data
```

# Figures

```{bash}
mkdir figures
```

```{r}
plotColors <- rev(RColorBrewer::brewer.pal(5, "Oranges")) #Create a color palette for the barplots. Use 5 purple shades from RColorBrewer. Reverse the order so the darkest shade is used first.
```

```{r}
head(sampleMetadata) #Look at sample metadata
```

## Comparative analysis

`methylKit` creates PCA biplots, but I want to create a custom biplot and multiplanel figure with the PCA and ANOTHER FIGURE to visualize the comparative analysis results.

### Modify global PCA

The correlation plot is difficult to modify, so I just need to modify the PCA.

```{r}
allDataPCA <- PCASamples(methylationInformationFilteredCov5, obj.return = TRUE) #Run a PCA analysis on percent methylation for all samples. methylKit uses prcomp to create the PCA matrix
summary(allDataPCA) #Look at summary statistics. The first PC explains 15.7% of variation, the second PC explains 14.8% of variation
```

```{r}
#pdf("figures/all-sample-PCA.pdf", width = 11, height = 8.5)

par(mar = c(5, 5, 1, 1)) #Specify inner and outer margins

fig.allDataPCA <- ordiplot(allDataPCA, choices = c(1, 2), type = "none", display = "sites", cex = 0.5, xlab = "", ylab = "", xaxt = "n", yaxt = "n") #Use ordiplot to create base biplot. Do not add any points
points(fig.allDataPCA, "sites", col = c(rep(plotColors[2], times = 4), rep(plotColors[4], times = 4)), pch = c(rep(16, times = 4), rep(17, times = 4)), cex = 3) #Add each sample. Darker samples are low pH, lighter samples are ambient

#Add multiple white boxes on top of the default black box to manually change the color
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")
box(col = "white")

ordiellipse(allDataPCA, sampleMetadata$pHTreatment, show.groups = "1", col = plotColors[4]) #Add confidence ellipse around the samples in ambient pH
ordiellipse(allDataPCA, sampleMetadata$pHTreatment, show.groups = "0", col = plotColors[2]) #Add confidence ellipse around the samples in low pH

axis(side =  1, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add x-axis
mtext(side = 1, text = "PC 1 (15.7%)", line = 3, cex = 1.5) #Add x-axis label

axis(side =  2, labels = TRUE, col = "grey80", cex.axis = 1.7) #Add y-axis
mtext(side = 2, text = "PC 2 (14.8%)", line = 3, cex = 1.5) #Add y-axis label

legend("bottomright", 
       pch = c(16, 17), 
       legend = c("Low pH", "Ambient pH"), 
       col = c(plotColors[2], plotColors[4]), 
       cex = 1.7, bty = "n") #Add a legend with information about ambient and elevated samples

#dev.off()
```

## Heatmap

I will use `heatmap.2` to create a multipanel figure with four heatmaps: female DML that are not SNPs, female DML-SNP overlaps, indeterminate DML that are not SNPs, and indeterminate DML-SNP overlaps. I will use `heatmap.2` over `pheatmap` because it allows me to add a density plot with a legend.

### Female samples

#### Format data

```{r}
DMLPositionsFem <- rep(0, times = length(diffMethStatsTreatment75Fem$chr)) #Create an empty vector to store row numbers
for (i in 1:length(DMLPositionsFem)) {
  DMLPositionsFem[i] <- which(getData(diffMethStatsTreatment75Fem)$start[i] == getData(methylationInformationFilteredCov5Fem)$start)
} #For each DML, save the row number where that DML is found in methylationInformationFilteredCov5Fem. This will be used to subset the methylBase object with DML information
tail(DMLPositionsFem) #Confirm vector was created
```

```{r}
DMLMatrixFem <- methylationInformationFilteredCov5Fem[DMLPositionsFem,] #Subset methylationInformationFilteredCov5Destrand to only include DML and save as a new methylBase object
sum((DMLMatrixFem$start) == (diffMethStatsTreatment75Fem$start)) == length(diffMethStatsTreatment75Fem$start) #Confirm that start columns are identical. If they are identical, the sum of all TRUE statements should equal the length of the original methylBase object
tail(DMLMatrixFem) #Confirm methylBase object creation
```

```{r}
percMethDMLFem <- methylKit::percMethylation(DMLMatrixFem, rowids = TRUE) #Get percent methylation for all samples at DML. Include row IDs (chr, start, end) information
head(percMethDMLFem) #Check that the DML matrix was created
```

```{r}
#Create a heatmap with all female DML data

#pdf("figures/all-female-heatmap.pdf", width = 11, height = 8.5)
par(mar = c(5, 12, 1, 1)) #Specify inner and outer margins

heatmap.2(percMethDMLFem, col = rev(plotColors), scale = "none", margins = c(1,1),
          trace = "none", tracecol = "black",
          labRow = FALSE, labCol = FALSE, 
          ColSideColors = c(rep("grey10", times = 3), rep("grey90", times = 3)),
          key = TRUE, keysize = 1.8, density.info = "density", key.title = "", key.xlab = "% Methylation", key.ylab = "",
          key.par = list(cex.lab = 2.0, cex.axis = 1.5)) #Create heatmap using heatmap.2 from gplots package using percMethDML data. Use plotColors but do not scale data, label rows, or label columns. Use ColSideColors to indicate colors for low (dark) and ambient (light) pH samples. Add a legend using key, and adjust keysize. Have key display density data with density.info. Do not add a key title or y-axis label, and label x axis with key.xlab.

#dev.off()
```

#### DML-SNP overlaps

```{r}
uniqueSNPFem <- read.table("../10_DML-characterization/DML-pH-75-Cov5-Fem-unique-CT-SNPs.bed", header = FALSE, sep = "\t", col.names = c("chr", "start", "end", "meth.diff")) #Import overlaps between unique C/T SNPs and female DML produced with bedtools
head(uniqueSNPFem) #Confirm import
```

```{r}
uniqueSNPPositionsFem <- rep(0, times = length(uniqueSNPFem$chr)) #Create an empty vector to store row numbers
for (i in 1:length(uniqueSNPPositionsFem)) {
  uniqueSNPPositionsFem[i] <- which(uniqueSNPFem$start[i] == getData(methylationInformationFilteredCov5Fem)$start)
} #For each DML, save the row number where that DML is found in methylationInformationFilteredCov5Fem. This will be used to subset the methylBase object with DML information
tail(uniqueSNPPositionsFem) #Confirm vector was created
```

```{r}
uniqueSNPMatrixFem <- methylationInformationFilteredCov5Fem[uniqueSNPPositionsFem,] #Subset methylationInformationFilteredCov5Fem to only include unique SNP overlaps and save as a new methylBase object
tail(uniqueSNPMatrixFem) #Confirm methylBase object creation
```

```{r}
percMethUniqueSNPFem <- methylKit::percMethylation(uniqueSNPMatrixFem, rowids = TRUE) #Get percent methylation for all samples at DML. Include row IDs (chr, start, end) information
head(percMethUniqueSNPFem) #Check that the DML matrix was created
```

```{r}
#Create a heatmap with DML-unique SNP overlap data

#pdf("figures/SNP-female-heatmap.pdf", width = 11, height = 8.5)
par(mar = c(5, 12, 1, 1)) #Specify inner and outer margins

heatmap.2(percMethUniqueSNPFem, col = rev(plotColors), scale = "none", margins = c(1,1),
          trace = "none", tracecol = "black",
          labRow = FALSE, labCol = FALSE, 
          ColSideColors = c(rep("grey10", times = 3), rep("grey90", times = 3)),
          key = TRUE, keysize = 1.8, density.info = "density", key.title = "", key.xlab = "% Methylation", key.ylab = "",
          key.par = list(cex.lab = 2.0, cex.axis = 1.5)) #Create heatmap using heatmap.2 from gplots package using percMethUniqueSNPFem data. Use plotColors but do not scale data, label rows, or label columns. Use ColSideColors to indicate colors for low (dark) and ambient (light) pH samples. Add a legend using key, and adjust keysize. Have key display density data with density.info. Do not add a key title or y-axis label, and label x axis with key.xlab.

#dev.off()
```

#### Non-SNP DML

```{r}
percMethUniqueDMLFem <- percMethDMLFem[which(!(rownames(percMethDMLFem) %in% rownames(percMethUniqueSNPFem))),] #Identify rownames from percMethDMLFem in percMethUniqueSNPFem (%in%), then print the opposite (!). Use which() to get row numbers for subsetting. 
head(percMethUniqueDMLFem) #Confirm 
```

```{r}
#Create a heatmap with DML-unique SNP overlap data

#pdf("figures/uniqueDML-female-heatmap.pdf", width = 11, height = 8.5)
par(mar = c(5, 12, 1, 1)) #Specify inner and outer margins

heatmap.2(percMethUniqueDMLFem, col = rev(plotColors), scale = "none", margins = c(1,1),
          trace = "none", tracecol = "black",
          labRow = FALSE, labCol = FALSE, 
          ColSideColors = c(rep("grey10", times = 3), rep("grey90", times = 3)),
          key = TRUE, keysize = 1.8, density.info = "density", key.title = "", key.xlab = "% Methylation", key.ylab = "",
          key.par = list(cex.lab = 2.0, cex.axis = 1.5)) #Create heatmap using heatmap.2 from gplots package using percMethUniqueDMLFem data. Use plotColors but do not scale data, label rows, or label columns. Use ColSideColors to indicate colors for low (dark) and ambient (light) pH samples. Add a legend using key, and adjust keysize. Have key display density data with density.info. Do not add a key title or y-axis label, and label x axis with key.xlab.

#dev.off()
```

### Indeterminate samples

#### Format data

```{r}
DMLPositionsInd <- rep(0, times = length(diffMethStatsTreatment100Ind$chr)) #Create an empty vector to store row numbers
for (i in 1:length(DMLPositionsInd)) {
  DMLPositionsInd[i] <- which(getData(diffMethStatsTreatment100Ind)$start[i] == getData(methylationInformationFilteredCov5Ind)$start)
} #For each DML, save the row number where that DML is found in methylationInformationFilteredCov5Ind. This will be used to subset the methylBase object with DML information
tail(DMLPositionsInd) #Confirm vector was created
```

```{r}
DMLMatrixInd <- methylationInformationFilteredCov5Ind[DMLPositionsInd,] #Subset methylationInformationFilteredCov5Destrand to only include DML and save as a new methylBase object
sum((DMLMatrixInd$start) == (diffMethStatsTreatment100Ind$start)) == length(diffMethStatsTreatment100Ind$start) #Confirm that start columns are identical. If they are identical, the sum of all TRUE statements should equal the length of the original methylBase object
tail(DMLMatrixInd) #Confirm methylBase object creation
```

```{r}
percMethDMLInd <- methylKit::percMethylation(DMLMatrixInd, rowids = TRUE) #Get percent methylation for all samples at DML. Include row IDs (chr, start, end) information
head(percMethDMLInd) #Check that the DML matrix was created
```

```{r}
#Create a heatmap with all indeterminate DML data

#pdf("figures/all-ind-heatmap.pdf", width = 11, height = 8.5)
par(mar = c(5, 12, 1, 1)) #Specify inner and outer margins

heatmap.2(percMethDMLInd, col = rev(plotColors), scale = "none", margins = c(1,1),
          trace = "none", tracecol = "black",
          labRow = FALSE, labCol = FALSE, 
          ColSideColors = c(rep("grey10", times = 1), rep("grey90", times = 1)),
          key = TRUE, keysize = 1.8, density.info = "density", key.title = "", key.xlab = "% Methylation", key.ylab = "",
          key.par = list(cex.lab = 2.0, cex.axis = 1.5)) #Create heatmap using heatmap.2 from gplots package using percMethDML data. Use plotColors but do not scale data, label rows, or label columns. Use ColSideColors to indicate colors for low (dark) and ambient (light) pH samples. Add a legend using key, and adjust keysize. Have key display density data with density.info. Do not add a key title or y-axis label, and label x axis with key.xlab.

#dev.off()
```

#### DML-SNP overlaps

```{r}
uniqueSNPInd <- read.table("../10_DML-characterization/DML-pH-100-Cov5-Ind-unique-CT-SNPs.bed", header = FALSE, sep = "\t", col.names = c("chr", "start", "end", "meth.diff")) #Import overlaps between unique C/T SNPs and female DML produced with bedtools
head(uniqueSNPInd) #Confirm import
```

```{r}
uniqueSNPPositionsInd <- rep(0, times = length(uniqueSNPInd$chr)) #Create an empty vector to store row numbers
for (i in 1:length(uniqueSNPPositionsInd)) {
  uniqueSNPPositionsInd[i] <- which(uniqueSNPInd$start[i] == getData(methylationInformationFilteredCov5Ind)$start)
} #For each DML, save the row number where that DML is found in methylationInformationFilteredCov5Ind. This will be used to subset the methylBase object with DML information
tail(uniqueSNPPositionsInd) #Confirm vector was created
```

```{r}
uniqueSNPMatrixInd <- methylationInformationFilteredCov5Ind[uniqueSNPPositionsInd,] #Subset methylationInformationFilteredCov5Ind to only include unique SNP overlaps and save as a new methylBase object
tail(uniqueSNPMatrixInd) #Confirm methylBase object creation
```

```{r}
percMethUniqueSNPInd <- methylKit::percMethylation(uniqueSNPMatrixInd, rowids = TRUE) #Get percent methylation for all samples at DML. Include row IDs (chr, start, end) information
head(percMethUniqueSNPInd) #Check that the DML matrix was created
```

```{r}
#Create a heatmap with DML-unique SNP overlap data

#pdf("figures/SNP-ind-heatmap.pdf", width = 11, height = 8.5)
par(mar = c(5, 12, 1, 1)) #Specify inner and outer margins

heatmap.2(percMethUniqueSNPInd, col = rev(plotColors), scale = "none", margins = c(1,1),
          trace = "none", tracecol = "black",
          labRow = FALSE, labCol = FALSE, 
          ColSideColors = c(rep("grey10", times = 1), rep("grey90", times = 1)),
          key = TRUE, keysize = 1.8, density.info = "density", key.title = "", key.xlab = "% Methylation", key.ylab = "",
          key.par = list(cex.lab = 2.0, cex.axis = 1.5)) #Create heatmap using heatmap.2 from gplots package using percMethUniqueSNPInd data. Use plotColors but do not scale data, label rows, or label columns. Use ColSideColors to indicate colors for low (dark) and ambient (light) pH samples. Add a legend using key, and adjust keysize. Have key display density data with density.info. Do not add a key title or y-axis label, and label x axis with key.xlab.

#dev.off()
```

#### Non-SNP DML

```{r}
percMethUniqueDMLInd <- percMethDMLInd[which(!(rownames(percMethDMLInd) %in% rownames(percMethUniqueSNPInd))),] #Identify rownames from percMethDMLInd in percMethUniqueSNPInd (%in%), then print the opposite (!). Use which() to get row numbers for subsetting. 
head(percMethUniqueDMLInd) #Confirm 
```

```{r}
#Create a heatmap with DML-unique SNP overlap data

#pdf("figures/uniqueDML-ind-heatmap.pdf", width = 11, height = 8.5)
par(mar = c(5, 12, 1, 1)) #Specify inner and outer margins

heatmap.2(percMethUniqueDMLInd, col = rev(plotColors), scale = "none", margins = c(1,1),
          trace = "none", tracecol = "black",
          labRow = FALSE, labCol = FALSE, 
          ColSideColors = c(rep("grey10", times = 1), rep("grey90", times = 1)),
          key = TRUE, keysize = 1.8, density.info = "density", key.title = "", key.xlab = "% Methylation", key.ylab = "",
          key.par = list(cex.lab = 2.0, cex.axis = 1.5)) #Create heatmap using heatmap.2 from gplots package using percMethUniqueDMLInd data. Use plotColors but do not scale data, label rows, or label columns. Use ColSideColors to indicate colors for low (dark) and ambient (light) pH samples. Add a legend using key, and adjust keysize. Have key display density data with density.info. Do not add a key title or y-axis label, and label x axis with key.xlab.

#dev.off()
```

`heatmap.2` calls a new plot each time so I cannot easily make a multipanel figure. I will do this in InDesign or PowerPoint.

## Chromosome distribution

I want to look at how DML (those that do and do not overlap with SNPs) are distributed across chromosomes of the genome.

```{r}
geneGFF <- read.table("../../genome-feature-files/cgigas_uk_roslin_v1_gene.gff", sep = "\t") #Import gene gff file
genechrCounts <-as.data.frame(table(geneGFF$V1)) #Count the number of genes/chromosome
colnames(genechrCounts) <- c("chr", "geneCount") #Rename columns
nrow(genechrCounts) #There are 10 chromosomes + extra scaffolds that were not placed into chromosomes
```

```{r}
CGmotifs <- read.table("../../genome-feature-files/cgigas_uk_roslin_v1_fuzznuc_CGmotif.gff", sep = "\t") #Import CG motif file
CGmotifsCounts <-as.data.frame(table(CGmotifs$V1)) #Count the number of CpGs/chromosome
colnames(CGmotifsCounts) <- c("chr", "chrCpGCounts") #Rename columns
nrow(CGmotifsCounts) #There are 10 chromosomes + extra scaffolds that were not placed into chromosomes
head(CGmotifsCounts)
```

### Female samples

```{r}
DMLchrCountsFem <-as.data.frame(table(diffMethStatsTreatment75Fem$chr)) #Count the number of DML/chromosome
colnames(DMLchrCountsFem) <- c("chr", "DMLCount") #Rename columns
head(DMLchrCountsFem) #Confirm formatting
```

```{r}
DMLchrCountsFem <- merge(x = DMLchrCountsFem, y = genechrCounts, by = "chr") #Merge by chr to get gene counts for each chr with DML
DMLchrCountsFem <- merge(x = DMLchrCountsFem, y = CGmotifsCounts, by = "chr") #Merge by chr to get CpG counts for each chr with DML
DMLchrCountsFem$DMLbyChrCpG <- (DMLchrCountsFem$DMLCount / DMLchrCountsFem$chrCpGCounts) #Normalize DML counts by number of CpGs in each chromosome

write.csv(DMLchrCountsFem, "DML/female-DML-by-chr.csv", row.names = TRUE, quote = FALSE) #Save file
```

```{r}
DMLchrCountsFem <- DMLchrCountsFem[1:10,] #Retain the first 10 linkage groups, since this is the data in chrosomomes and not unplaced scaffolds
```

### Indeterminate samples

```{r}
DMLchrCountsFem <-as.data.frame(table(diffMethStatsTreatment100Ind$chr)) #Count the number of DML/chromosome
colnames(DMLchrCountsFem) <- c("chr", "DMLCount") #Rename columns
head(DMLchrCountsFem) #Confirm formatting
```

```{r}
DMLchrCountsFem <- merge(x = DMLchrCountsFem, y = genechrCounts, by = "chr") #Merge by chr to get gene counts for each chr with DML
DMLchrCountsFem <- merge(x = DMLchrCountsFem, y = CGmotifsCounts, by = "chr") #Merge by chr to get CpG counts for each chr with DML
DMLchrCountsFem$DMLbyChrCpG <- (DMLchrCountsFem$DMLCount / DMLchrCountsFem$chrCpGCounts) #Normalize DML counts by number of CpGs in each chromosome

write.csv(DMLchrCountsFem, "DML/indeterminate-DML-by-chr.csv", row.names = TRUE, quote = FALSE) #Save file
```

```{r}
DMLchrCountsFem <- DMLchrCountsFem[1:10,] #Retain the first 10 linkage groups, since this is the data in chrosomomes and not unplaced scaffolds
```

### Create plot

```{r}
par(mar = c(5,7,2,10)) #Change figure boundaries

DMLbarplotFem <- barplot(as.matrix(t(DMLchrCountsFem$DMLbyChrCpG)),
                      axes = FALSE,
                      names.arg = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10"),
                      xlim = c(0.7,11.5),
                      ylim = c(0,5e-05),
                      cex.names = 1.5,
                      col = plotColors[4]) #Create a barplot and save as a new object. Use axes = FALSE to remove the y-axis and names.arg to set labels on the x-axis. The object contains x coordinates for bars, so xlim is set at 11.5
mtext(side = 1, "Chromosome", line = 3, cex = 1.5) #Add x-axis label
axis(side = 2, line = 1.5, at = seq(0, 5e-05, by = 1e-05), labels = seq(0, 5, by = 1), las = 2, col = "grey80", cex.axis = 1.2) #Add y-axis for DML counts
mtext(side = 2, "Number DML per 10,000 CpGs", line = 5, cex = 1.5) #Add y-axis label for DML counts

par(new = TRUE) #Create a new plot
plot(x = DMLbarplotFem,
     y = DMLchrCountsFem$geneCount,
     type = "b",
     axes = FALSE, xlab = "", ylab = "", xaxs = "i", yaxs = "i",
     pch = 16, col = plotColors[1],
     xlim = c(0,12), ylim = c(0,4500)) #Plot points and lines (type = "b") for gene count by chromosome. Use the coordinates from DMLbarplotFem (x = DMLbarplotFem) and set xlim = (0,12) so plots are lined up. Use axes = FALSE to remove both axes. Remove x and y lables (xlab = ""; ylab = ""). Set ylim = (0,4500) to account for max y-values. Use xaxs and yxaxs to remove space between axes.
axis(side = 4, line = 1.5, at = seq(0, 6500, by = 500), las = 2, col = "grey80", cex.axis = 1.2) #Add y-axis for gene sequence counts
mtext(side = 4, "Number of Genes", line = 6, cex = 1.5) #Add y-axis label for gene sequence counts
```



## Hypothetical gene distribution





