---
title: "Genomic Location of DML"
author: "Yaamini Venkataraman"
output: github_document
---

In this script, I'll create summary tables for genomic location information, run statistcal tests, and visualize the distribution of DML in the genome.

# Prepare R Markdown file

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/Users/yaaminivenkataraman/Documents/project-gigas-oa-meth/output/10_DML-characterization/") #Set root directory
```

# Install packages

```{r}
#Packages for visualization

#install.packages("dichromat")
#install.packages("broom")

require(dichromat)
require(broom)
```

# Obtain session information

```{r}
sessionInfo()
```

# Establish color scheme

```{r}
plotColors <- rev(RColorBrewer::brewer.pal(5, "Oranges")) #Create a color palette for the barplots. Use 5 purple shades from RColorBrewer. Reverse the order so the darkest shade is used first.
```

```{r}
plotColors2 <- rev(RColorBrewer::brewer.pal(8, "Oranges")) #Create a color palette for the barplots. Use 5 purple shades from RColorBrewer. Reverse the order so the darkest shade is used first.
```

# Import data

```{r}
#save.image("../../project-gigas-oa-meth.Rdata")
load("../../project-gigas-oa-meth.Rdata") #Load R Data if not already loaded
```

I will look at genomic locations of each sex-specific DML list, then create a plot for visualization.

# All samples

## Import file counts

```{r}
cpgFeatureOverlapsAll <- read.table("DML-pH-50-Cov5-All-Overlap-counts.txt", header = FALSE, sep = "\t", col.names = c("allDML", "filename")) #Import line counts
cpgFeatureOverlapsAll <- cpgFeatureOverlapsAll[-c(2, 3, 10, 12),] #Drop extra lines
cpgFeatureOverlapsAll <- cpgFeatureOverlapsAll[c(4, 1, 6, 8, 3, 7, 2, 5),] #Reorganize rows to match order: exon UTR, CDS, intron, upstream, downstream, lncRNA, TE, intergenic
tail(cpgFeatureOverlapsAll) #Confirm import
```

```{r}
cpgFeatureOverlapsAll$allCpGs <- cpgFeatureOverlaps$allCpGs #Add genomic CpG information
rownames(cpgFeatureOverlapsAll) <- row.names(cpgFeatureOverlaps) #Add rowname information
cpgFeatureOverlapsAll <- cpgFeatureOverlapsAll[,-2] #Drop filename column
head(cpgFeatureOverlapsAll)
```

## Contingency test

```{r}
cpgLocationStatTestAll <- data.frame(t(cpgFeatureOverlapsAll)) #Transpose for statistical testing
head(cpgLocationStatTestAll) #Confirm formatting
```

```{r}
CpGLocationGenomeDMLAll <- data.frame() #Create empty dataframe to store chi-squared results
for(i in 1:ncol(cpgLocationStatTestAll)) { #For each genome feature
  GenomeFeature <- cpgLocationStatTestAll[2,i] #Variable for # genome feature overlaps for methylated CpGs
  AllFeature <- cpgLocationStatTestAll[1,i] #Variable for # genome feature overlaps for all DML
  GenomeNotFeature <- sum(cpgLocationStatTestAll[2,-i]) #Variable for # other CpG types for methylated CpGs
  AllNotFeature <- sum(cpgLocationStatTestAll[1,-i]) #Variable for # other CpG types for all DML
  ct <- matrix(c(GenomeFeature, AllFeature, GenomeNotFeature, AllNotFeature), ncol = 2) #Create contingency table
  colnames(ct) <- c(as.character(colnames(cpgLocationStatTestAll[i])), paste0("Not", colnames(cpgLocationStatTestAll[i]))) #Assign column names: type, not type
  rownames(ct) <- c(as.character(row.names(cpgLocationStatTestAll)[c(2,1)])) #Assign row names: methylated CpGs, all DML
  print(ct) #Confirm table is correct
  ctResults <- data.frame(broom::tidy(chisq.test(ct))) #Create dataframe storing chi-sq stats results. Use broom::tidy to extract results from test output
  ctResults$GenomeFeature <- as.character(colnames(cpgLocationStatTestAll)[i]) #Add CpG type to results
  CpGLocationGenomeDMLAll <- rbind(CpGLocationGenomeDMLAll, ctResults) #Add test statistics to master table
}
```

```{r}
head(CpGLocationGenomeDMLAll)
```

```{r}
CpGLocationGenomeDMLAll$p.adj <- p.adjust(CpGLocationGenomeDMLAll$p.value, method = "fdr") #Correct p-value using FDR
range(CpGLocationGenomeDMLAll$p.adj) #Look at range of p-values
head(CpGLocationGenomeDMLAll) #Confirm changes
```

```{r}
write.csv(CpGLocationGenomeDMLAll, "CpG-location-statResults-All.txt", quote = FALSE, row.names = TRUE) #Save statistical output
```

# Female samples

## Import file counts

```{r}
cpgFeatureOverlapsFem <- read.table("DML-pH-75-Cov5-Fem-Overlap-counts.txt", header = FALSE, sep = "\t", col.names = c("femaleDML", "filename")) #Import line counts
cpgFeatureOverlapsFem <- cpgFeatureOverlapsFem[-c(2, 3, 10, 12),] #Drop extra lines
cpgFeatureOverlapsFem <- cpgFeatureOverlapsFem[c(4, 1, 6, 8, 3, 7, 2, 5),] #Reorganize rows to match order: exon UTR, CDS, intron, upstream, downstream, lncRNA, TE, intergenic
tail(cpgFeatureOverlapsFem) #Confirm import
```

```{r}
cpgFeatureOverlapsFem$Meth <- cpgFeatureOverlaps$Meth #Add methylated CpG information
rownames(cpgFeatureOverlapsFem) <- row.names(cpgFeatureOverlaps) #Add rowname information
cpgFeatureOverlapsFem <- cpgFeatureOverlapsFem[,-2] #Drop filename column
head(cpgFeatureOverlapsFem)
```

## Contingency test

```{r}
cpgLocationStatTestFem <- data.frame(t(cpgFeatureOverlapsFem)) #Transpose for statistical testing
head(cpgLocationStatTestFem) #Confirm formatting
```
```{r}
CpGLocationMethDMLFem <- data.frame() #Create empty dataframe to store chi-squared results
for(i in 1:ncol(cpgLocationStatTestFem)) { #For each genome feature
  MethFeature <- cpgLocationStatTestFem[1,i] #Variable for # genome feature overlaps for methylated CpGs
  FemFeature <- cpgLocationStatTestFem[2,i] #Variable for # genome feature overlaps for female DML
  MethNotFeature <- sum(cpgLocationStatTestFem[1,-i]) #Variable for # other CpG types for methylated CpGs
  FemNotFeature <- sum(cpgLocationStatTestFem[2,-i]) #Variable for # other CpG types for female DML
  ct <- matrix(c(MethFeature, FemFeature, MethNotFeature, FemNotFeature), ncol = 2) #Create contingency table
  colnames(ct) <- c(as.character(colnames(cpgLocationStatTestFem[i])), paste0("Not", colnames(cpgLocationStatTestFem[i]))) #Assign column names: type, not type
  rownames(ct) <- c(as.character(row.names(cpgLocationStatTestFem)[c(1,2)])) #Assign row names: methylated CpGs, female DML
  print(ct) #Confirm table is correct
  ctResults <- data.frame(broom::tidy(chisq.test(ct))) #Create dataframe storing chi-sq stats results. Use broom::tidy to extract results from test output
  ctResults$GenomeFeature <- as.character(colnames(cpgLocationStatTestFem)[i]) #Add CpG type to results
  CpGLocationMethDMLFem <- rbind(CpGLocationMethDMLFem, ctResults) #Add test statistics to master table
}
```

```{r}
head(CpGLocationMethDMLFem)
```

```{r}
CpGLocationMethDMLFem$p.adj <- p.adjust(CpGLocationMethDMLFem$p.value, method = "fdr") #Correct p-value using FDR
range(CpGLocationMethDMLFem$p.adj) #Look at range of p-values
head(CpGLocationMethDMLFem) #Confirm changes
```

```{r}
write.csv(CpGLocationMethDMLFem, "CpG-location-statResults-Fem.txt", quote = FALSE, row.names = TRUE) #Save statistical output
```

# Indeterminate samples

## Import file counts

```{r}
cpgFeatureOverlapsInd <- read.table("DML-pH-100-Cov5-Ind-Overlap-counts.txt", header = FALSE, sep = "\t", col.names = c("indeterminateDML", "filename")) #Import line counts
cpgFeatureOverlapsInd <- cpgFeatureOverlapsInd[-c(2, 3, 10, 12),] #Drop extra lines
cpgFeatureOverlapsInd <- cpgFeatureOverlapsInd[c(4, 1, 6, 8, 3, 7, 2, 5),] #Reorganize rows to match order: exon UTR, CDS, intron, upstream, downstream, lncRNA, TE, intergenic
tail(cpgFeatureOverlapsInd) #Confirm import
```

```{r}
cpgFeatureOverlapsInd$Meth <- cpgFeatureOverlaps$Meth #Add methylated CpG information
rownames(cpgFeatureOverlapsInd) <- row.names(cpgFeatureOverlaps) #Add rowname information
cpgFeatureOverlapsInd <- cpgFeatureOverlapsInd[,-2] #Drop filename column
head(cpgFeatureOverlapsInd)
```

## Contingency test

```{r}
cpgLocationStatTestInd <- data.frame(t(cpgFeatureOverlapsInd)) #Transpose for statistical testing
head(cpgLocationStatTestInd) #Confirm formatting
```

```{r}
CpGLocationMethDMLInd <- data.frame() #Create empty dataframe to store chi-squared results
for(i in 1:ncol(cpgLocationStatTestInd)) { #For each genome feature
  MethFeature <- cpgLocationStatTestInd[1,i] #Variable for # genome feature overlaps for methylated CpGs
  IndFeature <- cpgLocationStatTestInd[2,i] #Variable for # genome feature overlaps for indeterminate DML
  MethNotFeature <- sum(cpgLocationStatTestInd[1,-i]) #Variable for # other CpG types for methylated CpGs
  IndNotFeature <- sum(cpgLocationStatTestInd[2,-i]) #Variable for # other CpG types for indeterminate DML
  ct <- matrix(c(MethFeature, IndFeature, MethNotFeature, IndNotFeature), ncol = 2) #Create contingency table
  colnames(ct) <- c(as.character(colnames(cpgLocationStatTestInd[i])), paste0("Not", colnames(cpgLocationStatTestInd[i]))) #Assign column names: type, not type
  rownames(ct) <- c(as.character(row.names(cpgLocationStatTestInd)[c(1,2)])) #Assign row names: methylated CpGs, indeterminate DML
  print(ct) #Confirm table is correct
  ctResults <- data.frame(broom::tidy(chisq.test(ct))) #Create dataframe storing chi-sq stats results. Use broom::tidy to extract results from test output
  ctResults$GenomeFeature <- as.character(colnames(cpgLocationStatTestInd)[i]) #Add CpG type to results
  CpGLocationMethDMLInd <- rbind(CpGLocationMethDMLInd, ctResults) #Add test statistics to master table
}
```

```{r}
head(CpGLocationMethDMLInd)
```

```{r}
CpGLocationMethDMLInd$p.adj <- p.adjust(CpGLocationMethDMLInd$p.value, method = "fdr") #Correct p-value using FDR
range(CpGLocationMethDMLInd$p.adj) #Look at range of p-values
head(CpGLocationMethDMLInd) #Confirm changes
```

```{r}
write.csv(CpGLocationMethDMLInd, "CpG-location-statResults-Ind.txt", quote = FALSE, row.names = TRUE) #Save statistical output
```

# Create stacked barplot

## All DML

```{r}
cpgFeatureOverlapsPercentsAll <- cpgFeatureOverlapsAll #Duplicate dataframe
for (i in 1:length(cpgFeatureOverlapsAll)) {
  cpgFeatureOverlapsPercentsAll[,i] <- (cpgFeatureOverlapsPercentsAll[,i] / (sum(cpgFeatureOverlapsPercentsAll[,i]))) * 100
} #Divide every entry by sum of the column and multiply by 100 to get percentages. Do not include gene information
head(cpgFeatureOverlapsPercentsAll) #Check calculations
```

```{r}
#pdf("figures/DML-feature-overlaps.pdf", width = 11, height = 8.5)

par(mar = c(1,5,0,1), oma = c(3, 1, 1, 11)) #Change figure boundaries

barplot(t(t(cpgFeatureOverlapsPercentsAll[,c(2,1)])), 
        col= plotColors2, 
        axes = FALSE, 
        names.arg = c("High", "DML"), cex.names = 1.5,
        ylim = c(0, 110)) #Create base plot. Everything should be white. Do not plot axes. Include sequencing type as labels and set size. Set y-axis specs. 

axis(side = 2, at = seq(0, 100, by = 10), las = 2, col = "grey80", cex.axis = 1.2) #Add y-axis
mtext(side = 2, "% CpGs", line = 3, cex = 1.5) #Add y-axis label

par(fig = c(0, 1, 0, 1), oma = c(0, 0, 0, 0), mar = c(0, 0, 0, 0), new = TRUE) #Create new plot
plot(0, 0, type = "n", bty = "n", xaxt = "n", yaxt = "n") #Add new plot on top of current plot
legend(x = 0.57, y = 0.87, 
       xpd = TRUE,
       legend = c("Exon UTR", 
                  "CDS", 
                  "Introns", 
                  "Upstream Flank", 
                  "Downstream Flank", 
                  "lncRNA", 
                  "TE", 
                  "Intergenic"),
       pch = 22, 
        col = "black", 
        pt.bg = plotColors2,
       bty = "n",
       cex = 1.5, 
       x.intersp = 0.7, xjust = 0) #Place a legend in the top right of the figure with no box
text("Genome Feature", x = 0.785, y = 0.879, cex = 1.5) #Add legend title that is aligned with legend

#dev.off()
```

## Sex-specific DML

```{r}
cpgFeatureOverlapsPercentsFem <- cpgFeatureOverlapsFem #Duplicate dataframe
for (i in 1:length(cpgFeatureOverlapsFem)) {
  cpgFeatureOverlapsPercentsFem[,i] <- (cpgFeatureOverlapsPercentsFem[,i] / (sum(cpgFeatureOverlapsPercentsFem[,i]))) * 100
} #Divide every entry by sum of the column and multiply by 100 to get percentages. Do not include gene information
head(cpgFeatureOverlapsPercentsFem) #Check calculations
```

```{r}
cpgFeatureOverlapsPercentsInd <- cpgFeatureOverlapsInd #Duplicate dataframe
for (i in 1:length(cpgFeatureOverlapsInd)) {
  cpgFeatureOverlapsPercentsInd[,i] <- (cpgFeatureOverlapsPercentsInd[,i] / (sum(cpgFeatureOverlapsPercentsInd[,i]))) * 100
} #Divide every entry by sum of the column and multiply by 100 to get percentages. Do not include gene information
head(cpgFeatureOverlapsPercentsInd) #Check calculations
```

```{r}
cpgFeatureOverlapsPercentsBind <- cbind(cpgFeatureOverlapsPercentsFem, cpgFeatureOverlapsPercentsInd) #Combine dataframes
cpgFeatureOverlapsPercentsBind <- cpgFeatureOverlapsPercentsBind[,c(2,1,3)] #Reorganize and drop columns
head(cpgFeatureOverlapsPercentsBind)
```

```{r}
#pdf("figures/DML-feature-overlaps.pdf", width = 11, height = 8.5)

par(mar = c(1,5,0,1), oma = c(3, 1, 1, 11)) #Change figure boundaries

barplot(t(t(cpgFeatureOverlapsPercentsBind[,c(1:3)])), 
        col= plotColors2, 
        axes = FALSE, 
        names.arg = c("High", "Female", "Indeterminate"), cex.names = 1.5,
        ylim = c(0, 110)) #Create base plot. Everything should be white. Do not plot axes. Include sequencing type as labels and set size. Set y-axis specs. 

axis(side = 2, at = seq(0, 100, by = 10), las = 2, col = "grey80", cex.axis = 1.2) #Add y-axis
mtext(side = 2, "% CpGs", line = 3, cex = 1.5) #Add y-axis label

par(fig = c(0, 1, 0, 1), oma = c(0, 0, 0, 0), mar = c(0, 0, 0, 0), new = TRUE) #Create new plot
plot(0, 0, type = "n", bty = "n", xaxt = "n", yaxt = "n") #Add new plot on top of current plot
legend(x = 0.57, y = 0.87, 
       xpd = TRUE,
       legend = c("Exon UTR", 
                  "CDS", 
                  "Introns", 
                  "Upstream Flank", 
                  "Downstream Flank", 
                  "lncRNA", 
                  "TE", 
                  "Intergenic"),
       pch = 22, 
        col = "black", 
        pt.bg = plotColors2,
       bty = "n",
       cex = 1.5, 
       x.intersp = 0.7, xjust = 0) #Place a legend in the top right of the figure with no box
text("Genome Feature", x = 0.785, y = 0.879, cex = 1.5) #Add legend title that is aligned with legend

#dev.off()
```