---
title: "Chacterizing Methylation Islands"
output: github_document
---

In this script, I'll examine the distribution of various genome features within methylation islands for *C. gigas* and *C. virginica*. I'll also investigate the association between methylation island length and biological process GOterms for overlapping genes.

# Set up R Markdown Document

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Install packages

```{r}
#install.packages("dichromat")
require(dichromat)
```

# Session information

```{r}
sessionInfo()
```

# Genome feature distribution in methylation islands

## *C. gigas*

### Import and reformat data

#### Exons

```{r}
exonMI <- read.delim("../2019-09-15-DML-Analysis/2020-02-11-Exon-MI.txt", header = FALSE) #Import overlaps
exonMI <- exonMI[,-c(2:3, 6:10)] #Remove redundant columns
colnames(exonMI) <- c("chr", "exon.start", "exon.end", "MI.start", "MI.end", "overlap") #Add column names
head(exonMI) #Confirm reformatting
```

```{r}
exonMI$MI.length <- exonMI$MI.end - exonMI$MI.start #Calculate MI length
exonMI$percent <- exonMI$overlap / exonMI$MI.length #Calculate percent overlap
head(exonMI) #Check calculations
```

```{r}
boxplot(exonMI$percent)
```

#### Introns

```{r}
intronMI <- read.delim("../2019-09-15-DML-Analysis/2020-02-11-Intron-MI.txt", header = FALSE) #Import overlaps
intronMI <- intronMI[,-c(2:3, 6:10)] #Remove redundant columns
colnames(intronMI) <- c("chr", "intron.start", "intron.end", "MI.start", "MI.end", "overlap") #Add column names
head(intronMI) #Confirm reformatting
```

```{r}
intronMI$MI.length <- intronMI$MI.end - intronMI$MI.start #Calculate MI length
intronMI$percent <- intronMI$overlap / intronMI$MI.length #Calculate percent overlap
head(intronMI) #Check calculations
```

#### Genes

```{r}
genesMI <- read.delim("../2019-09-15-DML-Analysis/2020-02-11-Gene-MI.txt", header = FALSE) #Import overlaps
genesMI <- genesMI[,-c(2:3, 6:10)] #Remove redundant columns
colnames(genesMI) <- c("chr", "genes.start", "genes.end", "MI.start", "MI.end", "overlap") #Add column names
head(genesMI) #Confirm reformatting
```

```{r}
genesMI$MI.length <- genesMI$MI.end - genesMI$MI.start #Calculate MI length
genesMI$percent <- genesMI$overlap / genesMI$MI.length #Calculate percent overlap
head(genesMI) #Check calculations
```

#### TE

```{r}
transElemMI <- read.delim("../2019-09-15-DML-Analysis/2020-02-11-TE-MI.txt", header = FALSE) #Import overlaps
transElemMI <- transElemMI[,-c(2:3,6:10)] #Remove redundant columns
colnames(transElemMI) <- c("chr", "TE.start", "TE.end", "MI.start", "MI.end", "overlap") #Add column names
head(transElemMI) #Confirm reformatting
```

```{r}
transElemMI$MI.length <- transElemMI$MI.end - transElemMI$MI.start #Calculate MI length
transElemMI$percent <- transElemMI$overlap / transElemMI$MI.length #Calculate percent overlap
head(transElemMI) #Check calculations
```

### Create standalone boxplot

```{r}
boxplot(exonMI$percent, intronMI$percent, transElemMI$percent, 
        outline = FALSE, 
        lty = 1, lwd = 1.5,
        axes = FALSE, names = c("Exons", "Introns", "Transposable Elements"), ylim = c(0,0.8),
        col = plotColors[2:4], border = "grey20") #Create a boxplot with exon, intron, and transposable element MI overlap data that are colored according to plotColors. Do not plot outlier information (outline = FALSE). Do not use dashed lines (lty = 1), increase line thickness (lwd = 1.5), and change border color (border).

mtext(side = 1, "Genomic Feature", line = 2, cex = 1.5) #Add x-axis label

axis(side = 2, at = seq(from = 0, to = 0.8, by = 0.2), las = 2, col = "grey80", cex = 1.2) #Add y-axis
mtext(side = 2, "Overlap with MI (%)", cex = 1.5, line = 3) #Add y-axis label

legend("topleft",
       legend = c("Exons", "Introns", "Transposable Elements"),
       pch = 22, 
        col = "black", 
        pt.bg = plotColors[2:4],
       cex = 1.2, bty = "n") #Place a legend in the top left of the figure with no box
```

## *C. virginica* and *C. gigas*

### Chi-squared tests of homogeneity

#### Import and reformat data

```{r}
comboDataFeatures <- read.csv("2020-02-11-Common-MI-Feature-Overlap-Proportions.csv", header = TRUE)
rownames(comboDataFeatures) <- comboDataFeatures$genomicFeature #Set genomic feature indication as rownames
comboDataFeatures <- comboDataFeatures[,-1] #Remove genomic feature indication column
head(comboDataFeatures) #Confirm import
```

```{r}
comboPropDataFeatures <- comboDataFeatures #Copy overlap data as a new dataframe
comboFeaturesTotalEntries <- c(sum(comboPropDataFeatures$Cvirg),
                               sum(comboPropDataFeatures$Cgigas)) #Sum the number of CpGs in each category
comboPropDataFeatures <- rbind(comboPropDataFeatures, comboFeaturesTotalEntries) #Add comboTotalEntries to bottom of proportionData
row.names(comboPropDataFeatures)[4] <- "totalEntries" #Add rowname
head(comboPropDataFeatures) #Confirm changes
```

```{r}
nLength <- length(comboPropDataFeatures$Cvirg) #Count number of rows
for(i in 1:nLength) {
  comboPropDataFeatures[i,] <- (comboPropDataFeatures[i,]/comboPropDataFeatures[4,])*100
} #Divide each column of comboPropData by respective totalEntries entry. Multiply by 100 and and save the percentage
comboPropDataFeatures <- comboPropDataFeatures[-4,] #Remove totalEntries
head(comboPropDataFeatures) #Confirm changes
```

The null hypothesis is that various genomic features are distributed in the same proportions in MI across species.

```{r}
speciesFeaturesTest <- chisq.test(t(comboPropDataFeatures)) #Conduct a chi-squared test
speciesFeaturesTest #The distribution of features in MI is significantly different between species.
```

### Grouped barplot

#### Import and reformat *C. virginica* data

##### Exons

```{r}
CvexonMI <- read.delim("../../../yaamini-virginica/analyses/2018-11-01-DML-and-DMR-Analysis/2020-02-06-MI-Exons.txt", header = FALSE) #Import overlaps
CvexonMI <- CvexonMI[,-4] #Remove redundant column
colnames(CvexonMI) <- c("chr", "exon.start", "exon.end", "MI.start", "MI.end", "overlap") #Add column names
head(CvexonMI) #Confirm reformatting
```

```{r}
CvexonMI$MI.length <- CvexonMI$MI.end - CvexonMI$MI.start #Calculate MI length
CvexonMI$percent <- CvexonMI$overlap / CvexonMI$MI.length #Calculate percent overlap
head(CvexonMI) #Check calculations
```

##### Introns

```{r}
CvintronMI <- read.delim("../../../yaamini-virginica/analyses/2018-11-01-DML-and-DMR-Analysis/2020-02-06-MI-Introns.txt", header = FALSE) #Import overlaps
CvintronMI <- CvintronMI[,-4] #Remove redundant column
colnames(CvintronMI) <- c("chr", "intron.start", "intron.end", "MI.start", "MI.end", "overlap") #Add column names
head(CvintronMI) #Confirm reformatting
```

```{r}
CvintronMI$MI.length <- CvintronMI$MI.end - CvintronMI$MI.start #Calculate MI length
CvintronMI$percent <- CvintronMI$overlap / CvintronMI$MI.length #Calculate percent overlap
head(CvintronMI) #Check calculations
```

##### Genes

```{r}
CvgenesMI <- read.delim("../../../yaamini-virginica/analyses/2018-11-01-DML-and-DMR-Analysis/2020-02-06-MI-Genes.txt", header = FALSE) #Import overlaps
CvgenesMI <- CvgenesMI[,-4] #Remove redundant column
colnames(CvgenesMI) <- c("chr", "genes.start", "genes.end", "MI.start", "MI.end", "overlap") #Add column names
head(CvgenesMI) #Confirm reformatting
```

```{r}
CvgenesMI$MI.length <- CvgenesMI$MI.end - CvgenesMI$MI.start #Calculate MI length
CvgenesMI$percent <- CvgenesMI$overlap / CvgenesMI$MI.length #Calculate percent overlap
head(CvgenesMI) #Check calculations
```

##### TE

```{r}
CvtransElemMI <- read.delim("../../../yaamini-virginica/analyses/2018-11-01-DML-and-DMR-Analysis/2020-02-06-MI-TEall.txt", header = FALSE) #Import overlaps
CvtransElemMI <- CvtransElemMI[,-c(2:3,6:10)] #Remove redundant columns
colnames(CvtransElemMI) <- c("chr", "TE.start", "TE.end", "MI.start", "MI.end", "overlap") #Add column names
head(CvtransElemMI) #Confirm reformatting
```

```{r}
CvtransElemMI$MI.length <- CvtransElemMI$MI.end - CvtransElemMI$MI.start #Calculate MI length
CvtransElemMI$percent <- CvtransElemMI$overlap / CvtransElemMI$MI.length #Calculate percent overlap
head(CvtransElemMI) #Check calculations
```

#### Create figure

```{r}
plotColors2 <- rev(RColorBrewer::brewer.pal(5, "Oranges")) #Create a color palette for the barplots. Use 5 purple shades from RColorBrewer. Reverse the order so the darkest shade is used first.
```

```{r}
#pdf("2019-02-11-Cv-Cg-Methylation-Island-Boxplots.pdf", width = 11, height = 8.5) #Save as a pdf

par(mar = c(4, 5, 1, 7)) #Change figure boundaries

boxplot(CvexonMI$percent, CvintronMI$percent, CvtransElemMI$percent,
        exonMI$percent, intronMI$percent, transElemMI$percent,
        at = c(1, 1.85, 2.7, 5, 5.85, 6.7),
        outline = FALSE, lty = 1, lwd = 1.5,
        axes = FALSE, names = c("Exons", "Introns", "Transposable Elements"), ylim = c(0,0.8),
        col = plotColors2[2:4], border = "grey20") #Create a boxplot with exon, intron, and transposable element MI overlap data for both species that are colored according to plotColors. Group species information together using at. Do not plot outlier information (outline = FALSE). Do not use dashed lines (lty = 1), increase line thickness (lwd = 1.5), and change border color (border).

mtext(side = 1, expression(italic(" C. virginica                                                                        C. gigas     ")), line = 0.5, cex = 1.2) #Add species information
mtext(side = 1, "Genomic Feature", line = 2, cex = 1.5) #Add x-axis label

axis(side = 2, at = seq(from = 0, to = 0.8, by = 0.2), las = 2, col = "grey80", cex = 1.2) #Add y-axis
mtext(side = 2, "Proportion Overlap with MI", cex = 1.5, line = 3) #Add y-axis label

par(fig = c(0, 1, 0, 1), oma = c(0, 0, 0, 0), mar = c(0, 0, 0, 0), new = TRUE) #Create new plot
plot(0, 0, type = "n", bty = "n", xaxt = "n", yaxt = "n") #Add new plot on top of current plot
legend("topright",
       xpd = TRUE,
       legend = c("Exons", "Introns", "Transposable Elements"),
       pch = 22, 
        col = "black", 
        pt.bg = plotColors2[2:4],
       cex = 1.2, bty = "n") #Place a legend in the top left of the figure with no box

#dev.off()
```

# Methylation island length and associated biological processes for overlapping genes

## Import and reformat data

```{r}
DMLMI <- read.delim("../2019-09-15-DML-Analysis/2020-02-11-DML-MI.txt", header = FALSE) #Import overlaps between DML and MI
DMLMI <- DMLMI[,-c(4:5)] #Remove redundant columns
colnames(DMLMI) <- c("chr", "DML.start", "DML.end", "MI.start", "MI.end", "overlap") #Add column names
DMLMI$MI.start.end <- paste(DMLMI$MI.start, "-", DMLMI$MI.end, sep = "") #Combine start and end information for MI
head(DMLMI) #Confirm reformatting
```

```{r}
DMLMIOverlapCounts <- as.data.frame(table(DMLMI$MI.start.end)) #Count the frequency of each methylation island in MI-DML overlaps
colnames(DMLMIOverlapCounts) <- c("MI.start.end", "DML.overlaps") #Rename columns
head(DMLMIOverlapCounts) #Confirm counting
```

```{r}
geneIDMI <- read.delim("../2019-09-15-DML-Analysis/2020-02-11-Gene-MI.txt", header = FALSE) #Import overlaps
geneIDMI <- geneIDMI[,-c(2:3, 6:8, 10)] #Remove redundant columns, but include gene ID information
colnames(geneIDMI) <- c("chr", "genes.start", "genes.end", "geneID", "MI.start", "MI.end", "overlap") #Add column names
geneIDMI$geneID <- substr(geneIDMI$geneID, 4, 15) #Remove extraneous characters surrounding gene ID information
geneIDMI$length <- geneIDMI$MI.end - geneIDMI$MI.start #Add MI length information
geneIDMI$MI.start.end <- paste(geneIDMI$MI.start, "-", geneIDMI$MI.end, sep = "") #Combine start and end information for MI
head(geneIDMI) #Confirm reformatting
```

```{r}
BPGOSlim <- read.delim("../2019-09-15-DML-Analysis/Blastquery-GOslim-BP.sorted.unique.noOther", sep = "\t", header = FALSE) #Import unique BP GOslim terms without any "other biological processes" entries. File is tab-delimited and does not have a header.
colnames(BPGOSlim) <- c("geneID", "GOSlim") #Rename columns
head(BPGOSlim) #Confirm import
```

## Merge dataframes

I want my final dataframe to contain MI start and end information, gene start and end information, biological process GOterms, and the number of overlapping DML.

```{r}
genesBPMI <- merge(geneIDMI, BPGOSlim, by = "geneID") #Merge gene-MI overlaps and biological process GOSlim terms by geneID. Do not keep gene-MI overlaps without associated biological process information
head(genesBPMI) #Confirm merging
```

```{r}
genesBPMIDML <- merge(genesBPMI, DMLMIOverlapCounts, by = "MI.start.end", all.x = TRUE) #Merge gene-MI overlap and biological process information with number of DML overlaps for each MI. Keep MI that don't have associated DML overlap information
genesBPMIDML$DML.overlaps[is.na(genesBPMIDML$DML.overlaps)] <- 0 #Replace NA in DML overlap column with 0 (no DML overlaps)
genesBPMIDML <- genesBPMIDML[,c(3, 2, 4, 5, 6, 7, 8, 9, 10, 11)] #Rearrange columns
head(genesBPMIDML) #Confirm merging and reformatting
```

```{r}
write.table(genesBPMIDML, "2020-12-30-Gene-MI-Overlaps-BP-DMLCount.txt", sep = "\t", quote = FALSE, row.names = FALSE) #Save table with MI-gene overlaps, DML counts, and biological process GOSlim terms
```

## Determine length cut-offs

```{r}
methylationIslands <- read.delim("2020-02-11-Methylation-Islands-500_0.02_50-filtered.tab", header = FALSE) #Import methylation island track
colnames(methylationIslands) <- c("chr", "MI.start", "MI.end", "CpGs") #Rename columns
methylationIslands$length <- methylationIslands$MI.end - methylationIslands$MI.start #Calculate length
head(methylationIslands) #Confirm reformatting
```

```{r}
hist(methylationIslands$length) #Plot histogram of length. Most methylation islands are on the shorter end
median(methylationIslands$length) #Determine the median of length. This will be my cutoff for "short" (< median) and "long" (> median) methylation islands
```

```{r}
shortMI <- subset(genesBPMIDML, subset = genesBPMIDML$length <= median(methylationIslands$length)) #Subset information for short MI
range(shortMI$length) #Confirm subset
hist(shortMI$length) #Examine lengths represented
```

```{r}
longMI <- subset(genesBPMIDML, subset = genesBPMIDML$length > median(methylationIslands$length)) #Subset information for long MI
range(longMI$length) #Confirm subset
hist(longMI$length) #Examine lengths represented
```

## Examine characteristics associated with methylation island length

### DML

This should be relatively straightforward, with short MI having less overlapping DML than long MI. I'm not sure if any methylation island is going to have lots of DML.

```{r}
range(shortMI$DML.overlaps) #Number of DML associated with short MI
median(shortMI$DML.overlaps) #Median DML associated with short MI
median(shortMI$DML.overlaps[which(shortMI$DML.overlaps != 0)]) #Median DML overlaps excluding no overlaps
```

```{r}
range(longMI$DML.overlaps) #Number of DML associated with long MI
median(longMI$DML.overlaps) #Median DML associated with long MI
median(longMI$DML.overlaps[which(longMI$DML.overlaps != 0)]) #Median DML overlaps excluding no overlaps
```


```{r}
hist(shortMI$DML.overlaps) #Frequency distribution of DML per MI
hist(shortMI$DML.overlaps[which(shortMI$DML.overlaps != 0)]) #Frequency distribution of DML per MI excluding no overlaps
hist(longMI$DML.overlaps) #Frequency of DML per MI
hist(longMI$DML.overlaps[which(longMI$DML.overlaps != 0)]) #Frequency distribution of DML per MI excluding no overlaps
```

As expected, long MI had more DML than short MI, but both primarily had 1 DML overlapping (if there were any overlaps).


