---
title: "Chacterizing Methylation Islands"
output: github_document
---

# Set up R Markdown Document

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Install packages

```{r}
#install.packages("dichromat")
require(dichromat)
```

# Session information

```{r}
sessionInfo()
```

# *C. gigas*

## Import and reformat data

### Exons

```{r}
exonMI <- read.delim("../2019-09-15-DML-Analysis/2020-02-11-Exon-MI.txt", header = FALSE) #Import overlaps
exonMI <- exonMI[,-c(2:3, 6:10)] #Remove redundant columns
colnames(exonMI) <- c("chr", "exon.start", "exon.end", "MI.start", "MI.end", "overlap") #Add column names
head(exonMI) #Confirm reformatting
```

```{r}
exonMI$MI.length <- exonMI$MI.end - exonMI$MI.start #Calculate MI length
exonMI$percent <- exonMI$overlap / exonMI$MI.length #Calculate percent overlap
head(exonMI) #Check calculations
```

```{r}
boxplot(exonMI$percent)
```

### Introns

```{r}
intronMI <- read.delim("../2019-09-15-DML-Analysis/2020-02-11-Intron-MI.txt", header = FALSE) #Import overlaps
intronMI <- intronMI[,-c(2:3, 6:10)] #Remove redundant columns
colnames(intronMI) <- c("chr", "intron.start", "intron.end", "MI.start", "MI.end", "overlap") #Add column names
head(intronMI) #Confirm reformatting
```

```{r}
intronMI$MI.length <- intronMI$MI.end - intronMI$MI.start #Calculate MI length
intronMI$percent <- intronMI$overlap / intronMI$MI.length #Calculate percent overlap
head(intronMI) #Check calculations
```

### Genes

```{r}
genesMI <- read.delim("../2019-09-15-DML-Analysis/2020-02-11-Gene-MI.txt", header = FALSE) #Import overlaps
genesMI <- genesMI[,-c(2:3, 6:10)] #Remove redundant columns
colnames(genesMI) <- c("chr", "genes.start", "genes.end", "MI.start", "MI.end", "overlap") #Add column names
head(genesMI) #Confirm reformatting
```

```{r}
genesMI$MI.length <- genesMI$MI.end - genesMI$MI.start #Calculate MI length
genesMI$percent <- genesMI$overlap / genesMI$MI.length #Calculate percent overlap
head(genesMI) #Check calculations
```

### TE

```{r}
transElemMI <- read.delim("../2019-09-15-DML-Analysis/2020-02-11-TE-MI.txt", header = FALSE) #Import overlaps
transElemMI <- transElemMI[,-c(2:3,6:10)] #Remove redundant columns
colnames(transElemMI) <- c("chr", "TE.start", "TE.end", "MI.start", "MI.end", "overlap") #Add column names
head(transElemMI) #Confirm reformatting
```

```{r}
transElemMI$MI.length <- transElemMI$MI.end - transElemMI$MI.start #Calculate MI length
transElemMI$percent <- transElemMI$overlap / transElemMI$MI.length #Calculate percent overlap
head(transElemMI) #Check calculations
```

## Create standalone boxplot

```{r}
boxplot(exonMI$percent, intronMI$percent, transElemMI$percent, 
        outline = FALSE, 
        lty = 1, lwd = 1.5,
        axes = FALSE, names = c("Exons", "Introns", "Transposable Elements"), ylim = c(0,0.8),
        col = plotColors[2:4], border = "grey20") #Create a boxplot with exon, intron, and transposable element MI overlap data that are colored according to plotColors. Do not plot outlier information (outline = FALSE). Do not use dashed lines (lty = 1), increase line thickness (lwd = 1.5), and change border color (border).

mtext(side = 1, "Genomic Feature", line = 2, cex = 1.5) #Add x-axis label

axis(side = 2, at = seq(from = 0, to = 0.8, by = 0.2), las = 2, col = "grey80", cex = 1.2) #Add y-axis
mtext(side = 2, "Overlap with MI (%)", cex = 1.5, line = 3) #Add y-axis label

legend("topleft",
       legend = c("Exons", "Introns", "Transposable Elements"),
       pch = 22, 
        col = "black", 
        pt.bg = plotColors[2:4],
       cex = 1.2, bty = "n") #Place a legend in the top left of the figure with no box
```

# *C. virginica* and *C. gigas*

## Import and reformat *C. virginica* data

### Exons

```{r}
CvexonMI <- read.delim("../../../yaamini-virginica/analyses/2018-11-01-DML-and-DMR-Analysis/2020-02-06-MI-Exons.txt", header = FALSE) #Import overlaps
CvexonMI <- CvexonMI[,-4] #Remove redundant column
colnames(CvexonMI) <- c("chr", "exon.start", "exon.end", "MI.start", "MI.end", "overlap") #Add column names
head(CvexonMI) #Confirm reformatting
```

```{r}
CvexonMI$MI.length <- CvexonMI$MI.end - CvexonMI$MI.start #Calculate MI length
CvexonMI$percent <- CvexonMI$overlap / CvexonMI$MI.length #Calculate percent overlap
head(CvexonMI) #Check calculations
```

### Introns

```{r}
CvintronMI <- read.delim("../../../yaamini-virginica/analyses/2018-11-01-DML-and-DMR-Analysis/2020-02-06-MI-Introns.txt", header = FALSE) #Import overlaps
CvintronMI <- CvintronMI[,-4] #Remove redundant column
colnames(CvintronMI) <- c("chr", "intron.start", "intron.end", "MI.start", "MI.end", "overlap") #Add column names
head(CvintronMI) #Confirm reformatting
```

```{r}
CvintronMI$MI.length <- CvintronMI$MI.end - CvintronMI$MI.start #Calculate MI length
CvintronMI$percent <- CvintronMI$overlap / CvintronMI$MI.length #Calculate percent overlap
head(CvintronMI) #Check calculations
```

### Genes

```{r}
CvgenesMI <- read.delim("../../../yaamini-virginica/analyses/2018-11-01-DML-and-DMR-Analysis/2020-02-06-MI-Genes.txt", header = FALSE) #Import overlaps
CvgenesMI <- CvgenesMI[,-4] #Remove redundant column
colnames(CvgenesMI) <- c("chr", "genes.start", "genes.end", "MI.start", "MI.end", "overlap") #Add column names
head(CvgenesMI) #Confirm reformatting
```

```{r}
CvgenesMI$MI.length <- CvgenesMI$MI.end - CvgenesMI$MI.start #Calculate MI length
CvgenesMI$percent <- CvgenesMI$overlap / CvgenesMI$MI.length #Calculate percent overlap
head(CvgenesMI) #Check calculations
```

### TE

```{r}
CvtransElemMI <- read.delim("../../../yaamini-virginica/analyses/2018-11-01-DML-and-DMR-Analysis/2020-02-06-MI-TEall.txt", header = FALSE) #Import overlaps
CvtransElemMI <- CvtransElemMI[,-c(2:3,6:10)] #Remove redundant columns
colnames(CvtransElemMI) <- c("chr", "TE.start", "TE.end", "MI.start", "MI.end", "overlap") #Add column names
head(CvtransElemMI) #Confirm reformatting
```

```{r}
CvtransElemMI$MI.length <- CvtransElemMI$MI.end - CvtransElemMI$MI.start #Calculate MI length
CvtransElemMI$percent <- CvtransElemMI$overlap / CvtransElemMI$MI.length #Calculate percent overlap
head(CvtransElemMI) #Check calculations
```

## Create grouped boxplot

```{r}
plotColors2 <- rev(RColorBrewer::brewer.pal(5, "Oranges")) #Create a color palette for the barplots. Use 5 purple shades from RColorBrewer. Reverse the order so the darkest shade is used first.
```

```{r}
#pdf("2019-02-11-Cv-Cg-Methylation-Island-Boxplots.pdf", width = 11, height = 8.5) #Save as a pdf

par(mar = c(4, 5, 1, 7)) #Change figure boundaries

boxplot(CvexonMI$percent, CvintronMI$percent, CvtransElemMI$percent,
        exonMI$percent, intronMI$percent, transElemMI$percent,
        at = c(1, 1.85, 2.7, 5, 5.85, 6.7),
        outline = FALSE, lty = 1, lwd = 1.5,
        axes = FALSE, names = c("Exons", "Introns", "Transposable Elements"), ylim = c(0,0.8),
        col = plotColors2[2:4], border = "grey20") #Create a boxplot with exon, intron, and transposable element MI overlap data for both species that are colored according to plotColors. Group species information together using at. Do not plot outlier information (outline = FALSE). Do not use dashed lines (lty = 1), increase line thickness (lwd = 1.5), and change border color (border).

mtext(side = 1, expression(italic(" C. virginica                                                                        C. gigas     ")), line = 0.5, cex = 1.2) #Add species information
mtext(side = 1, "Genomic Feature", line = 2, cex = 1.5) #Add x-axis label

axis(side = 2, at = seq(from = 0, to = 0.8, by = 0.2), las = 2, col = "grey80", cex = 1.2) #Add y-axis
mtext(side = 2, "Proportion Overlap with MI", cex = 1.5, line = 3) #Add y-axis label

par(fig = c(0, 1, 0, 1), oma = c(0, 0, 0, 0), mar = c(0, 0, 0, 0), new = TRUE) #Create new plot
plot(0, 0, type = "n", bty = "n", xaxt = "n", yaxt = "n") #Add new plot on top of current plot
legend("topright",
       xpd = TRUE,
       legend = c("Exons", "Introns", "Transposable Elements"),
       pch = 22, 
        col = "black", 
        pt.bg = plotColors2[2:4],
       cex = 1.2, bty = "n") #Place a legend in the top left of the figure with no box

#dev.off()
```

